### DISCUSSION TITLE
How to visualize solar system with voltage, current and power?

### DISCUSSION THREAD

**Discussion by dhyananbho** [HomeSpan/HomeSpan#798] [Q&A] [answered]
Thank you for your great library with excellent documentation.

I would like to visualize a solar system in HomeKit and automatically trigger some actions depending on the state of the solar system. More precisely: display the voltage, current and power of the solar modules and the voltage, current and power to the consumers and not just a battery level in %. I measure these values with an INA226 chip.

I couldn't find a HomeKit service or characteristic to display voltage, current or power. Am I missing something or is there none? If there is no such service, what would be an easy way to display these values on an iPhone, iPad or Mac using ESP32 and HomeSpan?

Thank you for your comments and advice.

**Comment by HomeSpan** [comment, marked as answer]
Apple does not support Characteristics for voltage, current, or power, and it is unfortunately not possible to modify the Home App to add them.  However, HomeSpan supports the creation of Custom Services and Custom Characteristics, which can be used in other apps such as Eve for HomeKit (free on the App store) that do support these Characteristics.

See the [HomeSpan API](https://github.com/HomeSpan/HomeSpan/blob/master/docs/Reference.md#custom-characteristics-and-custom-services-macros) for details and examples on how to add Custom Characteristics and Services.  Though Eve does not publish the UUIDs for its proprietary Characteristics (presumably they want you to buy Eve HomeKit products) there are a [few sources](https://gist.github.com/simont77/3f4d4330fa55b83f8ca96388d9004e7d) on the web that have assembled some of the codes.
